File: leetcode/0704_binary_search.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
    int search(vector<int>& nums, int target) {
        int mid, first = 0, last = nums.size() - 1;

        while (first <= last) {
            mid = first + (last - first) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                last = mid - 1;
            } else {
                first = mid + 1;
            }
        }

        return -1;
    }
};

int main() {
    auto solution = Solution();

    vector<int> nums = {-1, 0, 3, 5, 9, 12};

    assert(4 == solution.search(nums, 9));
    assert(-1 == solution.search(nums, 2));
}

File: leetcode/0704_binary_search.py
# -*- coding: utf-8 -*-

from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        first, last = 0, len(nums) - 1
        while first <= last:
            mid = first + (last - first) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                last = mid - 1
            else:
                first = mid + 1
        return -1


if __name__ == '__main__':
    solution = Solution()

    assert 4 == solution.search([-1, 0, 3, 5, 9, 12], 9)
    assert -1 == solution.search([-1, 0, 3, 5, 9, 12], 2)
